cpp/validate_harvested_records.cc

```diff
@@ -39,274 +40,195 @@
-    for (const auto &field_info : combined_info) {
-        if (field_info.presence_ != ALWAYS)
-            continue;   // we only care about required fields that are missing
-
-        const auto equivalent_tag(EQUIVALENT_TAGS_MAP.find(field_info.name_));
-        if (seen_tags.find(field_info.name_) != seen_tags.end())
-            ;// required tag found
-        else if (equivalent_tag != EQUIVALENT_TAGS_MAP.end() and seen_tags.find(equivalent_tag->second) != seen_tags.end())
-            ;// equivalent tag found
-        else {
-            *error_message = "Record w/ control number " + record.getControlNumber() + " in \"" + journal_name
-                             + "\" is missing the always expected " + field_info.name_ + " field.";
-            meets_expectations = false;
-        }
-    }
 
-    return meets_expectations;
+void JournalSpecificValidator::findMissingTags(const std::string &journal_id, const std::set<std::string> &present_tags,
+                                               std::set<std::string> * const missing_tags) const
+{
+    const auto journal_id_and_validators(journal_ids_to_validators_map_.find(journal_id));
+    if (journal_id_and_validators == journal_ids_to_validators_map_.cend())
+        return;
+    journal_id_and_validators->second.findMissingTags(journal_id, present_tags, missing_tags);
 }
 
 
-void WriteToDatabase(DbConnection * const db_connection, const GeneralInfo &general_info, const JournalInfo &journal_info) {
-    for (const auto &field_info : journal_info) {
-        if (general_info.find(field_info.name_) == general_info.end())
-            db_connection->queryOrDie("INSERT INTO metadata_presence_tracer SET zeder_journal_id=(SELECT id FROM zeder_journals "
-                                      "WHERE zeder_id=" + db_connection->escapeAndQuoteString(journal_info.getZederId()) + " "
-                                      "AND zeder_instance=" + db_connection->escapeAndQuoteString(journal_info.getZederInstance()) + ")"
-                                      ", marc_field_tag=" + db_connection->escapeAndQuoteString(field_info.name_) +
-                                      ", field_presence='" + FieldPresenceToString(field_info.presence_) + "'");
+void LoadRules(DbConnection * const db_connection, GeneralValidator * const general_regular_article_validator,
+               JournalSpecificValidator * const journal_specific_regular_article_validator,
+               GeneralValidator * const general_review_article_validator,
+               JournalSpecificValidator * const journal_specific_review_article_validator)
+{
+    db_connection->queryOrDie(
+        "SELECT marc_field_tag,marc_subfield_code,field_presence,record_type FROM metadata_presence_tracer");
+    DbResultSet result_set(db_connection->getLastResultSet());
+    while (const auto row = result_set.getNextRow()) {
+        if (row["record_type"] == "regular_article") {
+            if (row.isNull("journal_id"))
```

Comment:
the `journal_id` column is not included in your sql statement

Reply:
... Also it might be useful to add `ORDER BY marc_field_tag, marc_subfield_code ASC` so that error messages will be in the correct order later

Reply:
Done

Reply:
You added `ORDER BY ...`, but `journal_id` is still missing

Reply:
Should be fixed now.